### PR TITLE
last refactoring before final release candidate

### LINKED ISSUE #11: General Representation of Cognate Sets (including partial cognacy)
the more I think about it and the more I talk about it with colleagues, it becomes evident that cognacy needs to expanded by the aspect of partial cognacy. In Indo-European, we have only spurious accounts, but in Tukano, Sino-Tibetan, and many other families, compounds are not just some little noise but constitute the basis of the lexical structure of the languages (30% in the nouns of Chinese dialects, but also pervasive in other families, like Tupi-Guarani).

Now, for partial cognacy, there is a straightforward representation. If we have words like "she-goat" and "he-goat", we can assign two IDs, that would render the stuff as "1 2" and "1 3". Note that we have cases in which we have the opposite order in the same language family, that is, we have "she-goat" in langauge A and "he-goat" in language B. Here, we would write "1 2" and "2 1". Note also that we _need_ to keep the information in one cell, since we want to store the order of the elements, and keep track of it for the alignmetns, which ideally would only be made for identical parts of the compounds. So we would have one alignment for cogid 1 and one for cogid 2, etc. In a file, we could represent this in a column by keeping the order of the original tokens, but adding a separator for morphemes:

``` text

TOKENS         COGID  ALIGNMENT           
sh e + g oa t       1 2         sh e + g oa t
h e + g oa t         1 3         h e + g oa t
g oa t + h e         3 1         g oa t + h e
```

So the alignment is no longer immediately visible when looking at the text-file, but automatically accessible and it can be displayed with tools that offer the service of comparing the cogid column with the column of segments (TOKENS). We loose transparency for those who want to read text-files, but we gain power for representing cognacy truthfully.

Furthermore, the format is consistent with unique-ID representations of cognacy:
- in a strict version, we require _identity_ of compound structure, which is nothing else than a check for identical strings in the cogid-representation which gives multiple numerical IDs in one cell separated by a space.
- in a loose version, we can make a connected component analysis of the underlying network and label the respective items as cognate, which share direct or indirect links with other items

My most recent idea was, that this should essentially be the basis, since it is so frequent in many language families, and also useful where it is less frequent (but never underestimate the degree of partial cognacy, also for IE languages, once you start being strict with morphemes!). And since those datasets where partial cognacy doesn't play a role would still simply show one ID for a cognate set, it would be largely consistent with the rest. 

But in danger of repeating myself here: we cannot ignore the problem of handling partial cognates. They are just too massively present in too many languages, and can also soon be fully handled by automatic cognate detection algorithms (provided, morpheme boundaries are known: I have recent highly promising results for lingpy).

### LINKED ISSUE #31: Add `macroarea` property to ontology
http://glottolog.org/meta/glossary#macroarea

### LINKED ISSUE #35: Specify that column names must be unique
mostly to simplify access, but also to enforce a common intuition.

### LINKED ISSUE #37: Provide URIs for all properties used in the CLDF standard in the Ontology
All properties used in tables specified by CLDF should have URIs in the CLDF Ontology - possibly suitable `subPropertyOf` existing properties in other vocabularies. This will make sure, our tools can recognize these properties unambiguously and vice versa other tools cannot mistake these properties.

This should include `ID`, forcing a format similar to the one for column names (in addition to adding a `primaryKey` property implicitely). To mark `ID`s as URIs, referencing entities defined outside of the dataset, an appropriate `aboutUrl` should be used.

### LINKED ISSUE #38: The *_source columns in the *CognateTables should be multivalued, too
Like the `Source` column in the `FormTable`, the `*_source` columns in `*CognateTable` should accept lists of references as values, too.

### PR DESCRIPTION
**Pull request by xrotwang**
Notable changes:
- The main spec is collapsed into one file, `README.md`.
- all columns in any default description have `propertyUrl`s from the ontology.
- in particular, foreign keys are modeled with explicit reference properties. The intended behaviour of `pycldf` (not yet implemented) will be to auto-generate corresponding foreign key specs when a reference property is encountered and the referenced table is present in a dataset.

Fixed issues:
- closes #37 
- closes #38 
- closes #31 
- closes #35 
- closes #11

### CODE REVIEW

**Comment by xrotwang** [comment]
@Anaphory please have a look at this PR

**Comment by xrotwang** [comment]
@Anaphory I also intend to enhance `pycldf` such that access to **any** tables, columns and foreign key constraints is mediated through the `pycldf.dataset.Dataset` object, and no recurse to the wrapped `TableGroup` is necessary. This should make it fairly easy to put together arbitrarily complex CLDF datasets.

File: components/borrowings/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+Thus, CLDF provides a borrowing table, with the default description
+[`BorrowingTable-metadata.json`](BorrowingTable-metadata.json).
+
+A `null` value for the `sourceFormReference` marks missing knowledge about the source word
+(or donor language). In this case, a `Comment` should be used to describe the available
+knowledge.
+
+Often datasets will add a column specifying reliability levels of the loan assessment
+(e.g. WOLD's ["borrowed score"](http://wold.clld.org/terms#borrowed_score)),
+but no common encoding scheme for such levels has been conventionalized yet.
```
Comment (Anaphory) [inline components/borrowings/README.md]: It would still be nice to give suggestions on how to eg. implement the WoLD assessments within the CLDF standard.

Comment (xrotwang) [reply components/borrowings/README.md]: I may do that as an example. But before I can put together examples, I have to update `pycldf`.

Comment (xrotwang) [reply components/borrowings/README.md]: I guess, it depends on the use case. If I wanted to stuff all of WOLD into CLDF - e.g. as export - I'd just go with a column modeling closely what I have in the database. If I wanted to use WOLD for automated cognate detection, I'd decide on a cut-off, e.g. consider everything with a borrowed score >= 0.5 as borrowed and don't code reliability at all.

Comment (Anaphory) [reply components/borrowings/README.md]: That's fine. I noticed it as a thing I would like to see, but then I saw that it should be possible to puzzle it together from the other components, which I think was not really possible before, so it's not urgent.

Comment (xrotwang) [reply components/borrowings/README.md]: Part of what I hope to achieve with CLDF is that such transformations between various CLDF representations are easy enough (be it with existing tools such as excel or csvkit, or with tools built on `pycldf`) for the linguist to do.

**Comment by Anaphory** [comment]
Hm, all the properties are very specific, and often required, so it looks very rigid.
This is certainly necessary to some extend, but I can spin out examples where I cannot build a CLDF dataset. Is there scope for supporting them?

- Take a cauteous cognate coder, who does not think in cognate classes but in pairwise correspondences. They would want to have a CognateTable, but specify “This form is related to that other form [in this manner]” (in a structure very much like a loan word table, but with different semantics). 
- If I specify loan words, I might not want to add the Donor-Concepts to my concepts table which may be chosen for a particular reason, and maybe also not include the form in the main Forms table. Or I might want to store the “Borrowing Score” not in the same CodeTable table as some values I have for words. Please clarify whether having several different instances of the same component type is permitted – pycldf as it currently stands does make it very hard.
- ValueTable is missing from the Components list.

**Comment by LinguList** [comment]
> Take a cauteous cognate coder, who does not think in cognate classes but in pairwise correspondences. They would want to have a CognateTable, but specify “This form is related to that other form [in this manner]” (in a structure very much like a loan word table, but with different semantics).

could you give an example? I don't get this one.

**Comment by xrotwang** [comment]
I guess the whole point of CLDF is to make automatic consumption possible. Thus, specificity wins.
For other models there's still the `Generic` module.

**Comment by Anaphory** [comment]
Ah, top, I missed that. Are there generic tables or columns, as well?

**Comment by Anaphory** [comment]
[Forms are true, argumentation is not:]

Alor-Besar (AN) `laˈraː` is clearly cognate, by systematic sound changes, to Lewoingu (AN) `ˈləraː`.
Lewoingu (AN) `ˈləraː` has a many r/dʒ correnspondences to Kedang (AN), so it looks reasonable that lɔdʒɔ might be a related form. I might tentatively assume that Wersing (TAP) `wudi` is related to that, which puts a lot of TAP words (eg. Blagar `ved`) with connections to that form in the same class. If all I can do is specify cognate classes, I either have a big class

    {laˈraː, ˈləraː, lɔdʒɔ, wudi, ved}

or two smaller classes 

    {laˈraː, ˈləraː, lɔdʒɔ, wudi, ved}

and I need to either decide on something or bloat my database with redundancy, and no-one will really be able to check the judgement.

If instead I specify the graph

    laˈraː → ˈləraː → lɔdʒɔ → wudi ← ved

with a reference for each of these connections, checking dubious connections and seeing how they change the cognate typology is easier.

The old Dyen database actually has a mixed approach, which I don't think CLDF can map, either: There are cognate classes, cognate classes that have doubtful connections to other cognate classes, cognate classes that have doubtful internal connections plus a doubtful connection to some other class, and sets of forms that independently of each other are doubtfully connected to some other class.

**Comment by LinguList** [comment]
Dyen is a nullege approach, as it links words without specifying the word forms. We can't retrieve the forms from their links, so we deliberately decided NEVER to look into that as an example for CLDF.

As to your example: you should work up a small toy example where you show that this specification can be handled and is useful, so we can see how to include it in CLDF. This is what I did with partial cognates.

I do not really see the point of your cognate typologies for now, as all boils down to the regularity of sound correspondences, which can only be assessed via alignments and should then be computed rather than encoded by a human. For alignments, it is trivial to add an extra column illustrating the certainty for each segment. There is a working example showing how this can be computed in lingpy, but before this comes to specification, it needs more testing...

**Comment by xrotwang** [comment]
I also think the importance of CLDF is in being sufficiently rigid to allow
tools to work with it automatically. When using the data requires a human
to assess the semantics of the fields and relations a well-defined
serialization format is cold comfort - although  the w3c spec underlying
CLDF will give you exactly that.

Am 11.10.2017 15:10 schrieb "Johann-Mattis List" <notifications@github.com>:

**Comment by xrotwang** [comment]
Regarding Dyen: The lexibank approach would be to encode
Dyen-as-it-was-used in a publication as CLDF. So the mutually incompatible
cognate sets would already have been resolved by the authors of the
publication.

Am 11.10.2017 15:18 schrieb "Robert Forkel" <xrotwang@googlemail.com>:

**Comment by LinguList** [comment]
Yes, with Dyen, but only if the data was also given in forms (not as a simple nexus, where the errors become really evident).

**Comment by Anaphory** [comment]
So it's as easy as
> Is there scope for supporting them?
“No, that would not be useful.”

That's fine. I think it would make sense to specify it somewhere explicitly, like adding this to the *Why?* section on the website and/or repeating parts of it in the readme:

### Why?
[…]
In order to decouple tools and data, it is important that the standard specifies not just the structure, but also the semantics of the data stored. New types of data cannot be immediately compatible with independently developed tools – which is what conformity would suggest –, so the CLDF specifications are very rigid on purpose and consider data with unspecified semantics invalid. For new data types, the CLDF standard will evolve to encompass them.

**Comment by xrotwang** [comment]
Good suggestion. I'd stress, though, that this evolution can happen within
CLDF, I.e. with having syntactically valid data - alas without specified
semantics.
I think I wrote a bit about this in the spec where modules are introduced.

Am 11.10.2017 15:44 schrieb "Gereon Kaiping" <notifications@github.com>:

**Comment by Anaphory** [comment]
Top! I had not seen it, but it's indeed in there explicitly:

> **While it is possible to add any number of CSV files to a CLDF dataset, the CLDF standard** recognizes (and attaches specified semantics) to tables described with a common property dc:conformsTo with one of the table type URIs of the CLDF ontology as value.

> Additionally, CLDF semantics **can** be assigned to individual columns by assigning one of the property URIs defined in the CLDF ontology as propertyUrl.

(emphasis mine). So, more like

Why?

[…]
In order to decouple tools and data, it is important that the standard specifies not just the structure, but also the semantics of the data stored. New types of data cannot be immediately compatible with independently developed tools – which is what conformity would suggest –, so the CLDF specifications are very rigid where possible, but can accomodate unspecified semantics where the standard has not yet evolved to encompass them.

**Comment by Anaphory** [comment]
In other words, I'm very happy with what the specs look like now!

File: components/cognates/README.md [hunk 2]
```diff
@@ -6,15 +6,8 @@ in analyzing the wordlist data, but also serves as intermediate step before
...
-- `Form_ID`: required; a foreign key to the `FormTable` of the wordlist
-- `Cognate_set_ID`: required; an identifier of the cognate set a form is judged to belong to
-- `Alignment`: optional; an [alignment](http://cldf.clld.org/v1.0/terms.rdf#alignment) of the segments of the form within its cognate set. Since alignments can make cognacy judgements more transparent and because they typically are somewhat expensive to compute, they should be stored and provided with a `CognateTable`
-- `Cognate_source`: [source specification](http://cldf.clld.org/v1.0/terms.rdf#source) listing sources of the cognacy judgement
-- `Alignment_source`: [source specification](http://cldf.clld.org/v1.0/terms.rdf#source) listing sources of the alignment
+is important, and therefore covered by CLDF. The default description of the cognate
+table is available in [`CognateTable-metadata.json`](CognateTable-metadata.json).
 
 It is recommended to add columns or other metadata describing the method used
 for the cognacy judgenments and the alignments, but no clear standard for these
```
Comment (LinguList) [inline components/cognates/README.md]: typo:
```
judgenments -> judgments
```

File: RELEASING.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+
+Releasing a new version of the CLDF spec
+========================================
+
+- Update CHANGELOG.md
+- Update all ontology URIs to the new version number
+  and run the tests `tests/test.py`
```
Comment (SimonGreenhill) [inline RELEASING.md]: Just "Run the tests `tests/test.py` ?

File: RELEASING.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+- Create a release of https://github.com/glottobank/cldf
+- Make sure the release is picked up by ZENODO
+- Copy the DOI from ZENODO and add it to the release notes
+- Copy terms.rdf to a new version directory in the gh-pages branch
+- push the gh-pages branch
```
Comment (SimonGreenhill) [inline RELEASING.md]: +- Push ...

File: components/codes/README.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+according to their strategy for reduplication into
+- Productive full and partial reduplication
+- Full reduplication only
+- No productive reduplication
+
+Specifying the full range of valid categories for a given parameter is useful for
+error checking, and also if subsets of the data do not exhibit all possibilities.
+This can be done by adding a codes table with the default description
+[`CodeTable-metadata.json`](CodeTable-metadata.json).
+
```
Comment (SimonGreenhill) [inline components/codes/README.md]: I like this a lot.

File: components/cognatesets/README.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+## Cognate sets
+
+In historical linguistics cognate judgements have often been postulated "en gros",
+i.e. as complete set, not based on justifications for invidual cognates. Thus, the
```
Comment (SimonGreenhill) [inline components/cognatesets/README.md]: en masse?

**Comment by SimonGreenhill** [comment]
Made some minor comments.

**Comment by SimonGreenhill** [comment]
This also fixes #11 right?

**Pull request merged**